EIPS/eip-5283.md

```diff
@@ -0,0 +1,112 @@
+
+### Alternatives
+
+There are alternative designs to implement RPGs on the EVM:
+
+1.	Transient storage opcodes (`TLOAD`/`TSTORE`) provide contract state that is kept between calls in the same transaction but it is not committed to the world state afterward. These opcodes also enable fine-grained parallelization. 
+
+2. An opcode `SSTORE_COUNT` that retrieves the number of `SSTORE` instructions executed. It enables also fine-grained execution parallelization, but `SSTORE_COUNT` is much more complex to use correctly as it returns the number `SSTORE` opcodes executed, not the number of reentrant calls. Reentrancy must be deducted from this value.
+
+3. A new `LOCKCALL` opcode that works similar to `STATICALL` but only blocks storage writes in the caller contract. This results in cheaper RPG, but it doesn't allow some contract functions to be free of the RPG.
+
+
+All these alternate proposals have the downside that they create new opcodes, and this is discuraged if the same functionality can be implemented with the same gas cost using precompiles. A new opcode requires modifying compilers, debuggers and static analysys tools.
+
+
+### Gas cost
+A gast cost of 100 represents a worst case resource consumption, which occurs when the stack is almost full (approximately 400 addresses) and it is fully scanned. As the stack is always present in RAM, the scanning is fast.  Note: Once code is implemented in geth, it can be benchmarked and the cost can be re-evaluated, as it may result to be lower in practice. As a precompile call currently costs 700 gas, the cost of stack scanning has a low impact on the total cost of the precompile call (800 gas in totall).
+
+The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](./eip-1283.md). Using the `Semaphore` precompile as a reentrancy check would currently cost 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost and therefore discourages its use, it is still much lower than the pre-EIP-1283 cost. If a reduction in precompile call cost is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use of the precompile-based RPG, it is suggested that this EIP is implemented together with a reduction in precompile calls cost.
+
+
+## Backwards Compatibility
+This change is a hard-fork and therefore all full nodes must be updated.
+
+## Test Cases
+
+```
+contract Test is ReentrancyGuard2 {
+    function second() external nonReentrant {
+    }
+    function first() external nonReentrant {
+        this.second();
+    }
+}
+```
+A call to second() directly from a transaction does not revert, but a call to first() does revert.
+
+
+## Security Considerations
+
+No security risks have been identified related to this change.
```

Comment:
```suggestion
Needs discussion.
```

Reply:
No security risk has been identified in recent discussions.